packages/mds-db/index.ts

```diff
@@ -537,6 +550,100 @@ async function readEvents(params: ReadEventsQueryParams): Promise<ReadEventsResu
   })
 }
 
+async function readHistoricalEvents(params: ReadHistoricalEventsQueryParams) {
+  const { provider_id: query_provider_id, end_time } = params
+  const client = await getReadOnlyClient()
+  const vals = new SqlVals()
+  const values = vals.values()
+  let sql = `SELECT      e2.provider_id,
+  e2.device_id,
+  e2.event_type,
+  e2.timestamp,
+  lat,
+  lng,
+  speed,
+  heading,
+  accuracy,
+  altitude,
+  recorded
+FROM
+(
+SELECT      provider_id,
+      device_id,
+      event_type,
+      timestamp
+FROM
+(
+SELECT      provider_id,
+          device_id,
+          event_type,
+          timestamp,
+          recorded,
+          RANK() OVER (PARTITION BY device_id ORDER BY timestamp DESC) AS rownum
+FROM        events
+WHERE       provider_id IS NOT NULL`
+  if (query_provider_id) {
+    sql += `AND         provider_id = '${query_provider_id}'`
+  }
+  sql += `\nAND         timestamp < '${end_time}'
+  ) e1
+  WHERE       rownum = 1
+  AND         event_type IN ('trip_enter',
+                       'trip_start',
+                       'trip_end',
+                       'reserve',
+                       'cancel_reservation',
+                       'provider_drop_off',
+                       'service_end',
+                       'service_start')
+  ) e2
+  INNER JOIN  telemetry
+  ON          e2.device_id = telemetry.device_id
+  AND         e2.timestamp = telemetry.timestamp
+  ORDER BY    provider_id,
+    device_id,
+    event_type`
+
+  const { rows } = await client.query(sql, values)
+  const events = rows.reduce((acc: VehicleEvent[], row) => {
+    const {
+      provider_id,
+      device_id,
+      event_type,
+      timestamp,
+      recorded,
+      lat,
+      lng,
+      speed,
+      heading,
+      accuracy,
+      altitude
+    } = row
+    acc.push({
```

Comment:
In general, I have a preference for immutable array functions. I try to avoid `.push`. It doesn't really matter here, but 
```
return acc.concat({...})
```
or
```
return [...acc, {...}]
```
would be immutable alternatives.